### ISSUE TITLE
assembling large number of loci (UCEs)

### ISSUE THREAD

**Issue by MareikeJaniak** [juliema/aTRAM#297]
Good morning, 

I was wondering if you could give me some advice for assembling a dataset of UCEs with a 600 bp flanking region. Many of the UCEs are assembling fairly quickly (some in ~10 min) but I'm running into issues with some loci taking a very long time to assemble (4+ hours) and others failing immediately with a "Bus error (core dumped)" message (this is on an HPC).

For the loci that are taking a very long time, the problem seems to be a very large number of blast hits, which then take a very long time to be written to the assembler input files:

2020-11-02 20:27:07 INFO : 9960 blast hits in iteration 2
2020-11-02 20:27:07 INFO : Writing assembler input files: iteration 2
2020-11-02 22:40:56 INFO : Assembling shards with velvet: iteration 2

I'm already limiting the number of blast hits with --max-target-seqs, but maybe I just need to limit them even further? Here's the command I'm running at the moment:

atram.py -b libraries/PD_0026 -q references/uce-seq/Node_${SLURM_ARRAY_TASK_ID}.fa --blast-word-size 20 --iterations 4 --max-target-seqs 60 --assembler=velvet --cpus 10 --velvet-exp-cov 8 --log-file logs/${SLURM_ARRAY_TASK_ID}.assemble.log --log-level debug -t tmp/ -o assembly_out/uce-test/PD_0026

This particular library has 166 shards, so 60x166 = 9960 blast hits is probably excessive? Other loci that are much quicker are assembling just fine with <200 blast hits, so should I just set --max-target-seqs to 5 or 10 to speed things up?

I'm running aTRAM as a job array on an HPC, with -q, rather than -Q, because I noticed that -Q was assembling each reference serially, rather than in parallel, so the loci that were taking a very long time were slowing things down a lot. I initially thought that the "Bus error (core dumped)" issues were being caused by running too many jobs that were accessing the same aTRAM library simultaneously, so I've tried limiting the number of jobs that run simultaneously to <10, but I'm still getting the error on ~10% of jobs (but always different ones). Have you encountered this problem before? 

Finally, I was wondering if the "auto" option is still available for setting the number of iterations? Many of the loci seem to assemble across the full length within 1-3 iterations, so it would be great to set the number of iterations to auto, rather than running 4 or 5 on all of them. 

I'd really appreciate any and all suggestions you might have! 

Best, 
Mareike

**Comment by juliema** [comment]
Hey Mareike,

Answers below.

On Tue, Nov 3, 2020 at 5:04 AM MareikeJaniak <notifications@github.com>
wrote:

> Good morning,
>
> I was wondering if you could give me some advice for assembling a dataset
> of UCEs with a 600 bp flanking region. Many of the UCEs are assembling
> fairly quickly (some in ~10 min) but I'm running into issues with some loci
> taking a very long time to assemble (4+ hours) and others failing
> immediately with a "Bus error (core dumped)" message (this is on an HPC).
>

We had this problem with UCEs as well and did restrict the max target seqs.

> For the loci that are taking a very long time, the problem seems to be a
> very large number of blast hits, which then take a very long time to be
> written to the assembler input files:
>
> 2020-11-02 20:27:07 INFO : 9960 blast hits in iteration 2
> 2020-11-02 20:27:07 INFO : Writing assembler input files: iteration 2
> 2020-11-02 22:40:56 INFO : Assembling shards with velvet: iteration 2
>
> I'm already limiting the number of blast hits with --max-target-seqs, but
> maybe I just need to limit them even further? Here's the command I'm
> running at the moment:
>
> atram.py -b libraries/PD_0026 -q
> references/uce-seq/Node_${SLURM_ARRAY_TASK_ID}.fa --blast-word-size 20
> --iterations 4 --max-target-seqs 60 --assembler=velvet --cpus 10
> --velvet-exp-cov 8 --log-file logs/${SLURM_ARRAY_TASK_ID}.assemble.log
> --log-level debug -t tmp/ -o assembly_out/uce-test/PD_0026
>

I would suggest using trinity or abyss for the assemblies they might do
better with those higher numbers.

> This particular library has 166 shards, so 60x166 = 9960 blast hits is
> probably excessive? Other loci that are much quicker are assembling just
> fine with <200 blast hits, so should I just set --max-target-seqs to 5 or
> 10 to speed things up?
>
Yes I would reduce it more.

> I'm running aTRAM as a job array on an HPC, with -q, rather than -Q,
> because I noticed that -Q was assembling each reference serially, rather
> than in parallel, so the loci that were taking a very long time were
> slowing things down a lot. I initially thought that the "Bus error (core
> dumped)" issues were being caused by running too many jobs that were
> accessing the same aTRAM library simultaneously, so I've tried limiting the
> number of jobs that run simultaneously to <10, but I'm still getting the
> error on ~10% of jobs (but always different ones). Have you encountered
> this problem before?
>
Can you send the log file that has one of these errors in it.


> Finally, I was wondering if the "auto" option is still available for
> setting the number of iterations? Many of the loci seem to assemble across
> the full length within 1-3 iterations, so it would be great to set the
> number of iterations to auto, rather than running 4 or 5 on all of them.
>

Use the -i option to limit the number of iteration.  We usually do 3 for
UCE loci, that will help speed things up.


> I'd really appreciate any and all suggestions you might have!
>
Good luck with this!

**Comment by MareikeJaniak** [comment]
Hi Julie, 

Thanks so much for the fast reply! I'll follow your suggestions for speeding things along. 

For the cases with the "Bus error (core dumped)," a log file is generated, but it is empty. The error is in the error file that gets generated by the slurm job script. Here is the full error file:

Lmod is automatically replacing "intel/2016.4" with "gcc/7.3.0".

-------------------------------------------------------------------------------
The following dependent module(s) are not currently loaded: gcccore/.5.4.0 (required by: icc/.2016.4.258, ifort/.2016.4.258)
-------------------------------------------------------------------------------

Due to MODULEPATH changes, the following have been reloaded:
  1) openmpi/2.1.1

The following have been reloaded with a version change:
  1) gcccore/.5.4.0 => gcccore/.7.3.0

/var/spool/slurmd/job54348148/slurm_script: line 17: 131687 Bus error               (core dumped) atram.py -b libraries/PD_0026 -q references/uce-seq/Node_${SLURM_ARRAY_TASK_ID}.fa --blast-word-size 20 --iterations 4 --max-target-seqs 60 --assembler=velvet --cpus 10 --velvet-exp-cov 8 --log-file logs/${SLURM_ARRAY_TASK_ID}.assemble.log --log-level debug -t tmp -o assembly_out/uce-test/PD_0026

Based on the error output, I initially thought that maybe the ${SLURM_ARRAY_TASK_ID} wasn't getting replaced correctly, but the script generates the log file with the correct variable name. The log file is empty, though, so it looks like the variable is being replaced correctly, but the job gets killed before atram starts running properly. In the past I've seen core dumped errors when there are memory issues, but the same job runs just fine when I start it in a separate array or by itself. So that led me to think that it was an issue with too many processes trying to access the same sqlite database? But that's just a guess.

**Comment by MareikeJaniak** [comment]
Hi Julie, 

I just wanted to give a quick update. The issues with the "Bus error (core dumped)" seem to stem from something with the SLURM job array. I tried running it again with gnu parallel and it worked without issues.

Decreasing --max-target-seqs further and setting the iterations to 3 also sped up the assembly step a lot and the majority of loci assembled in < 15 minutes. Velvet also worked very well, but I wanted to mention that I added the option "-cov_cutoff 3" to the velvetg command and it improved the assemblies quite a bit. 

Thanks again for your help and for putting aTRAM together, it's been a really useful tool for my work!
